### ISSUE TITLE
basic tutorial batch correction

### ISSUE THREAD

**Issue by adamgayoso** [YosefLab/scVI#498]
I think the tutorial is still a bit unclear. How do you let scVI correct batches for any dataset? If you dataset has non-zero batch indices will it automatically correct batches?

**Comment by JuHey** [comment]
Indeed. If one works with a .h5ad dataset, where would scvi look for the batch information? Would it look for 'batch_indices' in an adata.obs['batch_indices'], and what would it recognize there? I am trying to get this to work by adapting the 'interaction with scanpy' notebook but I am not seeing scvi reading the batch information. Any suggestions?

**Comment by adamgayoso** [comment]
https://github.com/YosefLab/scVI/blob/05920d1f85daa362d4fb694e588ab090bc84e207/scvi/dataset/anndataset.py#L28

You can tell the dataset constructor which key has batch information. Note these should be integers. To be safe you could just add an obs key called `batch_indices` that has ints and that will be read by default.

**Comment by JuHey** [comment]
Thank you for your explanation and suggestions. When adding `batch_indices` as an obs key to my dataset and defining it in the function call `scviDataset = AnnDatasetFromAnndata(adata, batch_label = "batch_indices")`, the batches are recognized correctly. However, I then get the following WARNING: `- scvi.dataset.dataset | This dataset has some empty cells, this might fail scVI inference.Data should be filtered with my_dataset.filter_cells_by_count()`, which is a bit strange as I had pre-filtered the dataset in scanpy and QC did not show any empty cells. And another error at the point when writing `adata.obsm["X_scvi"] = scvi_latent`with: `ValueError: Value passed for key 'X_scvi' is of incorrect shape. Values of obsm must match dimensions (0,) of parent. value had shape (6244, 6) while it should have had (14693,).` That is obviously the result of the downsampling of the dataset when processing and writing it to the `scviDataset` in the function call listed above. How would I extract myself out of this snag?

**Comment by adamgayoso** [comment]
This can only occur when some of the cells in `adata.X` are 0 (i.e., some entries of `adata.X.sum(axis=1)` are 0. Maybe there is some weird edge case - is the smallest number in batch_indices 0 (it should be)?

Also, is there a reason you're only using 6 latent dimensions?

**Comment by JuHey** [comment]
Thank you for your comments. I looked at this some more: the issues I observed went away by restarting jupyter lab. With that I could test higher latent dimensions and for the specific dataset, I settled on values between 10 and 20. I just tried the notebook on another dataset and I ran into the following (reproducible and not kernel-issue): at `scviDataset = AnnDatasetFromAnnData(adata_orginal, batch_label = 'batch_indices)`, I get an AttributeError: `'NoneType' object has not attribute 'shape'`. I think it has something to do with `initialize_self_attribute` in `dataset.py`. The dataset processes fine with other means. How can I troubleshoot this?
